### PR TITLE
Section for pluggable auth backend support

### LINKED ISSUE #5: Suggestion for addition: Pluggable HTTP(S) authentication provider
From pypa/pip#4475 and pypa/twine#362. Quoting myself:

> Both pip and twine use `requests` internally, so it might be a good idea to build an entrypoints-based plugin system that can be used by both. I expect corporations would be the main users as well, so it makes sense to ask them for resources.

### PR DESCRIPTION
**Pull request by uranusjr**
Close #5.

**What is the current situation/context?**
Standard packaging tools currently only supports package indexes using basic authentication.

**What ought to be fixed, made, or implemented?**
See pypa/pip#4475 and pypa/twine#362. A shared interface and implementation for various alternative authentication method support can be developed for both tools (and maybe more), so organisations can choose to install them to be able to use e.g. Kerberos to secure their private package indexes.

**What kinds of work are necessary to make this happen?**

* A survey of various auth methods, and how they can be implemented as a pluggable library.
* Develop an interface that tools (e.g. pip) can implement to detect authentication method support, and call into the library that provides it.
* Develop and maintain libraries that implement the various auth methods for people to install when the support is needed.

### CODE REVIEW

**Comment by pradyunsg** [comment]
With something like this, we'd also be able to drop the keyring out-of-the-box integration in those tools (at least in pip) in favor of a plugin like this -- which would help usability as well -- and is worth noting here.

**Comment by xmunoz** [comment]
Can you add these issues into the markdown file?

**Comment by brainwane** [comment]
I second @xmunoz -- I think you should add the details you shared in the comment on this pull request into the FUNDABLES item itself.

**Comment by uranusjr** [comment]
I’ve added the content of the PR description as a part of the patch.

File: FUNDABLES.md [hunk 1]
```diff
@@ -237,6 +237,17 @@ security checks, DevOps/infrastructure/platform services (including
 API/email integration), user experience work, technical writing for end
 users, project management, and community outreach.
 
+### Architecture to support alternative authentication methods in packaging tools
+
+Python packaging tools that interacts with package indexes, such as pip ([pypa/pip#4475](https://github.com/pypa/pip/issues/4475)) and twine ([pypa/twine#362](https://github.com/pypa/twine/issues/362)), currently only have simple authentication support to secure private sources, such as [basic access authentication](https://en.wikipedia.org/wiki/Basic_access_authentication). Open source tool maintainers acknowledge that stronger authentication methods, such as single sign-on, are sometimes required due to organisation policies. We believe it’s beneficial to develop a pluggable Python library that can be depended by the packaging tools to provide additional authentication methods, but lack both the use case and domain knowledge in the area. We are looking for funding and expertise support from organisations.
```
Comment (di) [inline FUNDABLES.md:242]:
```suggestion
Python packaging tools that interact with package indexes, such as pip ([pypa/pip#4475](https://github.com/pypa/pip/issues/4475)) and twine ([pypa/twine#362](https://github.com/pypa/twine/issues/362)), currently only have simple authentication support to secure private sources, such as [basic access authentication](https://en.wikipedia.org/wiki/Basic_access_authentication). Open source tool maintainers acknowledge that when using third-party indices, stronger authentication methods, such as single sign-on, are sometimes required due to organisation policies. We believe it’s beneficial to develop a pluggable Python library that can be depended by the packaging tools to provide additional authentication methods, but lack both the use case and domain knowledge in the area. We are looking for funding and expertise support from organisations.
```

File: FUNDABLES.md [hunk 2]
```diff
@@ -237,6 +237,17 @@ security checks, DevOps/infrastructure/platform services (including
...
+
+We are looking into developing a shared interface and implementation for various alternative authentication method support can be developed for both tools (and maybe more), so organisations can choose to install them to be able to use e.g. Kerberos to secure their private package indexes. The work involved would include the following topics:
```
Comment (di) [inline FUNDABLES.md:244]:
```suggestion
We are interested in developing a shared interface and implementation for various alternative authentication methods. Support can be developed for both tools (and maybe more), so organisations can choose to install them to be able to use e.g. Kerberos to secure their private package indexes. The work involved would include the following topics:
```

File: FUNDABLES.md [hunk 3]
```diff
@@ -237,6 +237,17 @@ security checks, DevOps/infrastructure/platform services (including
...
+
+* A survey of various auth methods, and how they can be implemented as a pluggable library.
```
Comment (di) [inline FUNDABLES.md:246]:
```suggestion
* A survey of various authentication methods, and how they can be implemented as a pluggable library.
```

File: FUNDABLES.md [hunk 4]
```diff
@@ -237,6 +237,17 @@ security checks, DevOps/infrastructure/platform services (including
...
+* Develop an interface that tools (e.g. pip) can implement to detect authentication method support, and call into the library that provides it.
+* Develop and maintain libraries that implement the various auth methods for people to install when the support is needed.
```
Comment (di) [inline FUNDABLES.md:248]:
```suggestion
* Develop and maintain libraries that implement the various auth methods for users to install when support is needed.
```

**Comment by uranusjr** [comment]
Thanks! Updated.

**Comment by brainwane** [comment]
Thanks @uranusjr and thanks reviewers! I have polished wording slightly in 5f62a3a00cb9492a6e52ac8685c008ac741e3c0f .

**Pull request merged**